src/include/nfs/toposort.ycp

```diff
@@ -0,0 +1,125 @@
+/**
+ * File:
+ *   toposort.ycp
+ *
+ * Module:
+ *   System Services (Runlevel) (formerly known as Runlevel Editor)
+ *
+ * Summary:
+ *   Topological sorting for script dependencies
+ *
+ * Authors:
+ *   Martin Vidner <mvidner@suse.cz>
+ *
+ * $Id$
+ */
+
+{
+    /**
+     * Topologically sort a directed acyclic graph, ie. linearize a
+     * partial ordering.
+     * (what if the graph is a multigraph??)
+     * @param g A DAG as a map:
+     *  nodes are keys, values are lists of nodes that are reached
+     *  by an edge from the respective key.
+     * @return [out, rest]<br>
+     *  out: a list containing the keys of the map in topological order<br>
+     *  rest: a list, empty if the graph was acyclic, otherwise it is
+     *   a superset of the nodes forming the cycle
+     *   and "out" is a partial result
+     */
+    define list< list<string> > TopologicalSort (map<string, list<string> > g) ``{
+	list<string> out = [];
+	map<string, integer> in_degree = (map<string, integer>) mapmap (string vertex, list<string> targets, g, ``($[vertex: 0]));
+        foreach (string vertex, list<string> targets, g, ``{
+	    foreach (string target, targets, ``{
+		in_degree[target] = in_degree[target]:0 + 1;
+	    });
+	});
+
+	// cycle
+	while (size (in_degree) > 0)
+	{
+	    // get the vertices that can go next because they have zero in degree
+	    map<string, integer> next_m = (map<string, integer>) filter (string vertex, integer d, in_degree, ``(d == 0));
+	    if (size (next_m) == 0)
+	    {
+		// the graph is cyclic!
+		break;
+	    }
+	    foreach (string vertex, integer dummy, next_m, ``{
+		out[size(out)] = vertex;
```

Comment:
Actually I think it was an attempt to trick the type system of an older dialect of YCP. Using `add` now.